### ISSUE TITLE
Jenkins pipeline compatibility

### ISSUE THREAD

**Issue by mehiel** [jenkinsci/saltstack-plugin#33]
Greetings,
 kudos for this great plugin. Any plans to support the pipeline architecture? Or any example usage out there, you can point me to?

It is clear that Jenkins will favor the pipeline arch so it's maybe something to consider for this plugin as well. I've also noticed that it's not listed in the [plugin compatibility](https://github.com/jenkinsci/pipeline-plugin/blob/master/COMPATIBILITY.md) list. Thnx for the support!

**Comment by webmailcontatos** [comment]
I'm very curious for the answer

**Comment by mchugh19** [comment]
Hey all. Sorry for the delay. I'm currently in the process of moving from the US to Ireland so I've got a bit less free time than usual.

As far as pipeline support is concerned, there is a "pipeline" development branch for that. However, it will likely require some cleanup that was started by Eric Smacking: https://github.com/ericsmalling/saltstack-plugin/tree/refactoring

Based on the previously mentioned international move, should anyone wish to help out, I'm all for it! ;)

**Comment by mchugh19** [comment]
Merged into master.

Pipelines syntax should be easily discoverable from the pipeline syntax popout including in newer versions of the pipeline plugin/jenkins you can use the new shortname as follows:

```
salt arguments: 'ls -la', authtype: 'pam', clientInterface: local(blockbuild: false, jobPollTime: 5, target: 'master', targetType: 'glob'), credentialsId: 'a3d814c2-84ed-4752-94a8-271791bb5375', function: 'cmd.run', kwarguments: 'runas=jenkins', servername: 'http://localhost:8000', target: 'master', targettype: 'glob'
```

Or if you want to use pipeline input it will work as follows:
```
def userInput
stage('prompt') {
  userInput = input message: 'Change parameters?', parameters: [
    string(defaultValue: 'master', description: 'What machines to target with command', name: 'target'),    
    string(defaultValue: 'runas=jenkins', description: 'Any kwargs', name: 'kwarguments')
    ]
  echo ("target is: "+userInput['target'])
}
  
stage('execute salt') {
  node {
    echo ("kwarguments is: "+userInput['kwarguments'])
    salt arguments: 'ls -la', authtype: 'pam', clientInterface: local(blockbuild: false, jobPollTime: 5, target: userInput['target'], targetType: 'glob'), credentialsId: 'a3d814c2-84ed-4752-94a8-271791bb5375', function: 'cmd.run', kwarguments: userInput['kwarguments'], servername: 'http://localhost:8000', target: 'master', targettype: 'glob'
  }
}
```


Before release I'd like to try and get output saving functional. So something like: 
```
def saltoutput = salt arguments: 'ls -la', authtype: 'pam', clientInterface: local(blockbuild: false, jobPollTime: 5, target: 'master', targetType: 'glob'), credentialsId: 'a3d814c2-84ed-4752-94a8-271791bb5375', function: 'cmd.run', kwarguments: 'runas=jenkins', servername: 'http://localhost:8000', target: 'master', targettype: 'glob'
echo ("Salt outputted ${saltoutput}")
```

**Comment by mchugh19** [comment]
This support was just extended to use a proper Jenkins Step. This means the salt-api output must now be saved into a variable, and is available for further processing or validation. 

```
saltresult = salt authtype: 'pam', clientInterface: local(arguments: '"sleep 5; ls -la"', blockbuild: true, function: 'cmd.run', target: 'master', targetType: 'list'), credentialsId: 'a3d814c2-84ed-4752-94a8-271791bb5375', servername: 'http://localhost:8000'
println(saltresult)
```

The syntax remains similar to the previous version, but since argument input is now (like the salt cli) parsed on space instead of the former comma, note the example argument quoting. Pipeline output is only exported in json. If the salt return is not saved into a variable and printed, then no results will be displayed in the Jenkins console.

**Comment by mchugh19** [comment]
Released 2.0

Printing results from within pipeline requires saving and printing output. 

```
import groovy.json.*

node() {
  saltresult = salt authtype: 'pam', clientInterface: local(arguments: '"ls -la"', blockbuild: true, function: 'cmd.run', jobPollTime: 16, target: '*', targetType: 'glob'), credentialsId: 'a3d814c2-84ed-4752-94a8-271791bb5375', servername: 'http://localhost:8000'
  def prettyJson = JsonOutput.prettyPrint(saltresult)
  println(prettyJson)
}
```
